File: backend/app/models/stats.py
"""Pydantic models for user stats and dashboard data."""

from datetime import datetime

from pydantic import BaseModel


class COREStageStats(BaseModel):
    stage: str  # "CONNECT" | "OBSERVE" | "RECOMMEND" | "EXECUTE"
    mastery_pct: float  # average stage score across last 30 days (0–100)
    delta: float | None = None  # change vs previous 30 days; None if no prior data


class RecentSessionActivity(BaseModel):
    session_id: str
    persona_name: str | None = None
    started_at: datetime
    score: int | None = None
    grade: str | None = None


class UserStatsResponse(BaseModel):
    current_streak: int
    longest_streak: int
    sessions_this_week: int
    weekly_goal: int
    avg_score_this_week: float | None = None
    core_mastery: list[COREStageStats]
    recent_activity: list[RecentSessionActivity]
    total_sessions: int

File: frontend/src/types/stats.ts
export interface COREStageStats {
  stage: 'CONNECT' | 'OBSERVE' | 'RECOMMEND' | 'EXECUTE';
  mastery_pct: number;
  delta: number | null;
}

export interface RecentSessionActivity {
  session_id: string;
  persona_name: string | null;
  started_at: string; // ISO datetime string
  score: number | null;
  grade: string | null;
}

export interface UserStatsResponse {
  current_streak: number;
  longest_streak: number;
  sessions_this_week: number;
  weekly_goal: number;
  avg_score_this_week: number | null;
  core_mastery: COREStageStats[];
  recent_activity: RecentSessionActivity[];
  total_sessions: number;
}
